### ISSUE TITLE
Add events endpoint for node deployment

### ISSUE THREAD

**Issue by zreigz** [kubermatic/kubermatic#2616]
**User Story**
During cluster creation, the user should be able to see all failures from the machine controller. Right now user can only check the machine controller logs to get a better understanding of what went wrong. Such a situation causes problems like this: #2511

The solution to this issue is to create a new endpoint:
`
/projects/{project_id}/dc/{dc}/clusters/{cluster_id}/nodedeployments/{nodedeployment_id}/nodes/events
`
This endpoint will list all events for given node deployment.
It can be easily introduced to the kubermatic dashboard to show node deployment failures.

Example of event with error
```
  kind: Event
  lastTimestamp: 2019-01-14T12:07:00Z
  message: 'Instance creation failed: trying to create a VM: compute.VirtualMachinesClient#CreateOrUpdate:
    Failure sending request: StatusCode=0 -- Original Error: autorest/azure: Service
    returned an error. Status=<nil> Code="SkuNotAvailable" Message="The requested
    size for resource ''/subscriptions/528474a9-8cc6-46c8-9b74-1dbc31c4bb7f/resourceGroups/cluster-dsfff66dfn/providers/Microsoft.Compute/virtualMachines/machinedeployment-kubermatic-dsfff66dfn-ffjnl-b65759bd7-mw9b2''
    is currently not available in location ''westeurope'' zones '''' for subscription
    ''528474a9-8cc6-46c8-9b74-1dbc31c4bb7f''. Please try another size or deploy to
    a different location or zones. See https://aka.ms/azureskunotavailable for details."'
  metadata:
    creationTimestamp: 2019-01-14T09:42:16Z
    name: machinedeployment-kubermatic-dsfff66dfn-ffjnl-b65759bd7-mw9b2.1579adb44b015680
    namespace: kube-system
    resourceVersion: "15101"
    selfLink: /api/v1/namespaces/kube-system/events/machinedeployment-kubermatic-dsfff66dfn-ffjnl-b65759bd7-mw9b2.1579adb44b015680
    uid: ade1b91a-17e0-11e9-a61e-ba96388ef6aa
  reason: CreateInstanceFailed
  reportingComponent: ""
  reportingInstance: ""
  source:
    component: machine-controller
  type: Warning

```

**Acceptance criteria**
- endpoint returns all events for node deployment
- a user can filter events by type
type:
```
/?warning=true
```
- all errors from machine controller during crating cluster are displayed in dasboard

**Comment by zreigz** [comment]
cc @maciaszczykm @p0lyn0mial @alvaroaleman @mrIncompetent

**Comment by alvaroaleman** [comment]
The approach we have been using for this kind of issue (errors in the machine-controller) has been by setting "Terminal Errors" in the machine-controller.

In contrest to that approach what is outlined here has the advantage that it will work even for errors we didn't anticipate. We shouldn't have both thought.

@p0lyn0mial WDYT?

**Comment by p0lyn0mial** [comment]
I agree, displaying events and errors seems to be a good idea. 

If possible I'd like to start simple and ideally don't add any filters on API or just one for example returning only warnings.

**Comment by zreigz** [comment]
Filtering just for events types: "Normal" or "Warning" is a good idea. I will create separate issues for backend and frontend implementation. Thanks for your feedback.

**Comment by mrIncompetent** [comment]
Directly translating `Warning` events as errors in the UI can be misleading as those events can also be transient errors which might be things like:
- Timeout talking to the provider (Or any other network related error)
- Failed to update the machine object

As the machine-controller retries everything we might show errors in the UI even though we successfully created a VM during the second iteration.
Avoiding this could involve a smart deduplication logic which checks if a specific `Warning` event has been emitted multiple times, which could help.

Going exclusively with terminal errors (`machine.status.errorReason` and `machine.status.errorMessage`) has the disadvantage that we might not identify every error correctly as terminal - Like we did in the past:
- https://github.com/kubermatic/machine-controller/pull/441
- https://github.com/kubermatic/machine-controller/pull/369

IMO:
**From a developers perspective**: 
We should go with events entirely and completely drop the terminal error code. Identifying terminal errors is a tedious and error prone work which does not grant us any benifit over what we can get from events.

**From a UX perspective**: 
We lose information we would show on the machine object which might be helpful during debugging.
On the other side: `kubectl describe machine` seems a better way to debug anyway.

**tl;dr** I would opt for only using events. Its simpler and more effective in the long run.

**Comment by zreigz** [comment]
I was talking with @maciaszczykm about using events in UI. It will be a separate panel for machine deployment events. The user will get a whole view of events. In the case when machines can't be created or it takes a lot of time I think it can be very helpful and it can be first place when end user can find some tips about the status of cluster  We are not going to transfer `warning` events to UI errors because it can be really misleading.
@mrIncompetent thanks for your feedback.

**Comment by mrIncompetent** [comment]
Do we actually have events on the `MachineDeployment` or just on the `Machine` ?
I remember that upstream(MachineSet & MachineDeployment controllers) don't emit any events at all.

**Comment by zreigz** [comment]
I noticed the kubermatic machine controller generates the events for `Machine` and `MachineDeployment`. The events are available in the user cluster.

**Comment by alvaroaleman** [comment]
There are not much cases in which MachineSet or MachineDeployment controllers could emit useful events, the only I can think of is if a webhook prevents them from creating MachineSets/Machines

**Comment by p0lyn0mial** [comment]
In general I think that errors and events complement each other, where events could help understand why a particular error has happened or could be used to implement a progress indicator for an end user.

The fact that MachineSet/Deployment don't emit useful events can change in the future.

Perhaps in this particular case #2511 we could change machine-controller to set an error on a resource.

**Comment by zreigz** [comment]
@p0lyn0mial @alvaroaleman @mrIncompetent
I've checked machine controller source code. We have only events for `machine`. There is no events for `machinedeployment`
![events.png](https://images.zenhubusercontent.com/5bdd9c8645b0290c78fb640f/a6610b5f-7825-4d11-9fdd-59c396c4b361)
